apps/ejabberd/src/mongoose_cassandra_sup.erl

```
@@ -0,0 +1,98 @@
+-export([register_worker/2,
+         select_worker/2,
+         get_all_workers/0,
+         get_all_workers/1]).
+
+list_pools() ->
+    Children = supervisor:which_children(ejabberd_sup),
+    [PoolName || {_Name={mongoose_cassandra_sup, PoolName}, _Pid, supervisor, _} <- Children].
+
+stop() ->
+    [stop(PoolName) || PoolName <- list_pools()].
+
+start(PoolName, Config) ->
+    create_worker_pool(PoolName),
+    supervisor:start_child(ejabberd_sup, supervisor_spec(PoolName, Config)).
+
+stop(PoolName) ->
+    Tag = {mongoose_cassandra_sup, PoolName},
+    supervisor:terminate_child(ejabberd_sup, Tag),
+    supervisor:delete_child(ejabberd_sup, Tag),
+    delete_worker_pool(PoolName).
+
+start_link(PoolName, Config) ->
+    supervisor2:start_link({local, ?MODULE}, ?MODULE, [PoolName, Config]).
+
+supervisor_spec(PoolName, Config) ->
+    {{mongoose_cassandra_sup, PoolName},
+     {?MODULE, start_link, [PoolName, Config]},
+     permanent,
+     infinity,
+     supervisor,
+     [?MODULE]}.
+
+worker_spec(PoolName, Addr, Port, WorkerNumber, ClientOptions) ->
+    {{PoolName, Addr, Port, WorkerNumber},
+     {mongoose_cassandra_worker, start_link, [PoolName, Addr, Port, ClientOptions]},
+     {permanent, 10}, %% Delay is 10 seconds
+     infinity,
+     worker,
+     [mongoose_cassandra_worker]}.
+
+worker_specs(PoolName, Servers, ClientOptions) ->
+    [worker_spec(PoolName, Addr, Port, WorkerNumber, ClientOptions)
+     || {Addr, Port, WorkerCount} <- Servers,
+        WorkerNumber <- lists:seq(1, WorkerCount)].
+
+init([PoolName, ClientOptions]) ->
+    Servers = proplists:get_value(servers, ClientOptions),
+    Specs = worker_specs(PoolName, Servers, ClientOptions),
+    {ok, {{one_for_one, 10, 1}, Specs}}.
+
+
+%%====================================================================
+%% Registration
+%%====================================================================
+
+create_worker_pool(PoolName) ->
+    pg2:create(group_name(PoolName)).
```

I'd rather see `poolboy` here (or sth else) to fail fast if there is no spare workers.